### PR TITLE
Add BeitBackbone

### PR DESCRIPTION
**Pull request by NielsRogge**
# What does this PR do?

This PR is part one of the bigger #25799. It adds the `BeitBackbone` class, making it possible to use this vision transformer as a backbone for downstream tasks, like depth estimation. 

To do:

- [x] make `out_indices` backwards compatible
- [x] check whether this class is compatible with Mask R-CNN

### CODE REVIEW

File: src/transformers/models/beit/configuration_beit.py [hunk 1]
```diff
@@ -139,14 +150,16 @@ def __init__(
...
+        semantic_out_indices=[3, 5, 7, 11],
         pool_scales=[1, 2, 3, 6],
         use_auxiliary_head=True,
         auxiliary_loss_weight=0.4,
         auxiliary_channels=256,
         auxiliary_num_convs=1,
         auxiliary_concat_input=False,
         semantic_loss_ignore_index=255,
+        out_features=None,
+        out_indices=None,
```
Comment (amyeroberts) [inline src/transformers/models/beit/configuration_beit.py:162]: We can't make this breaking change - there's no warning for the users the `out_indices` isn't being used for the same thing anymore and old configs will load setting the backbone `out_indices` to whatever it was in the saved config

File: src/transformers/models/data2vec/modeling_data2vec_vision.py [hunk 2]
```diff
@@ -1180,7 +1180,9 @@ def forward(
 
         # only keep certain features, and reshape
         # note that we do +1 as the encoder_hidden_states also includes the initial embeddings
-        features = [feature for idx, feature in enumerate(encoder_hidden_states) if idx + 1 in self.config.out_indices]
+        features = [
+            feature for idx, feature in enumerate(encoder_hidden_states) if idx + 1 in self.config.semantic_out_indices
+        ]
```
Comment (amyeroberts) [inline src/transformers/models/data2vec/modeling_data2vec_vision.py:1185]: We can't make this breaking change - it completely changes the behaviour if someone did the following: 

```
model = Data2VecVisionForSemanticSegmentation("facebook/data2vec-vision-base-ft1k", output_indices=(-2, -1))
```

**Review by amyeroberts — COMMENTED** [review]
This can't be merged as-is - the renaming of `out_indices` to `semantic_out_indices` is a large breaking change and needs to be handled properly to maintain backwards compatibility and no unexpected behaviours

File: src/transformers/models/beit/configuration_beit.py [hunk 3]
```diff
@@ -183,6 +200,26 @@ def __init__(
         self.auxiliary_concat_input = auxiliary_concat_input
         self.semantic_loss_ignore_index = semantic_loss_ignore_index
 
+        # handle backwards compatibility
+        if "segmentation_indices" in kwargs:
+            logger.warning(
+                "The `segmentation_indices` argument is deprecated and will be removed in a future version, use `out_indices` instead.",
+                FutureWarning,
+            )
+            out_indices = kwargs.pop("segmentation_indices")
```
Comment (amyeroberts) [inline src/transformers/models/beit/configuration_beit.py:209]: Why would segmentation_indices be passed in as a kwarg? It doesn't seem to be an existing argument for the config file

File: src/transformers/models/beit/configuration_beit.py [hunk 4]
```diff
@@ -183,6 +200,26 @@ def __init__(
...
+        elif out_indices == [3, 5, 7, 11]:
+            out_features = ["stage4", "stage6", "stage8", "stage12"]
+        elif out_indices == [7, 11, 15, 23]:
+            out_features = ["stage8", "stage12", "stage16", "stage24"]
```
Comment (amyeroberts) [inline src/transformers/models/beit/configuration_beit.py:213]: This hard coding isn't needed. Given stage_names and out_features you can directly get out_indices i.e.:

```suggestion
out_features, out_indices = get_aligned_output_features_output_indices(
    out_features=None, out_indices=out_indices, stage_names=stage_names
)
```

File: src/transformers/models/beit/configuration_beit.py [hunk 5]
```diff
@@ -100,6 +99,22 @@ class BeitConfig(PretrainedConfig):
...
+        out_features (`List[str]`, *optional*):
+            If used as backbone, list of features to output. Can be any of `"stem"`, `"stage1"`, `"stage2"`, etc.
+            (depending on how many stages the model has). If unset and `out_indices` is set, will default to the
+            corresponding stages. If unset and `out_indices` is unset, will default to the last stage.
+        out_indices (`List[int]`, *optional*):
+            If used as backbone, list of indices of features to output. Can be any of 0, 1, 2, etc. (depending on how
+            many stages the model has). If unset and `out_features` is set, will default to the corresponding stages.
+            If unset and `out_features` is unset, will default to the last stage.
+
+            Can also indicate the indices of the feature maps to use for semantic segmentation.
```
Comment (amyeroberts) [inline src/transformers/models/beit/configuration_beit.py:111]: No no no no no. Having flags with conditional meanings is not a good design.

Moreover, the mixin will update `out_indices` if the user tries to set `out_features` but there's no control on whether or not this is being used as a backbone, leading to potentially unexpected behaviour. 

As a start, this is how I would try to handle this: 

* Remove `out_indices` as an input argument to `__init__`
* Add `semantic_out_indices` as a new argument with the default `None`

* If `semantic_out_indices` is set and `out_indices` is passed in as a kwarg, raise an exception
* If `semantic_out_indices` is `None` and `out_indices` is passed in as a kwarg, set `semantic_out_indices` to `out_indices` 
* If `semantic_out_indices` is `None` and `out_indices` isn't in the kwargs, the `semantic_out_indices` defaults to  `[3, 5, 7, 9]`
* Update model config to not save `out_indices` out including a comment explaining why

You'll need to add tests to make sure that the config and model are properly created with fully backwards compatibility with old configs, as well as fully configurable when creating a model `from_pretrained` and kwargs e.g. `BeitBackbone.from_pretrained(checkpoint, out_indices=(-2, -1))` or  `BeitBackbone.from_pretrained(checkpoint, semantic_out_indices=(-2, -1))`.

**Review by amyeroberts — COMMENTED** [review]
The current handling of `out_indices` isn't safe or backwards compatible. I've added a comment with a suggestion on how to handle this.

File: src/transformers/models/beit/configuration_beit.py (hunk 3 above)
Comment (NielsRogge) [reply src/transformers/models/beit/configuration_beit.py:207]: The legacy models on the hub only use the `segmentation_indices` argument: https://huggingface.co/microsoft/beit-base-finetuned-ade-640-640/blob/main/config.json#L327-L332. Hence to make this PR backwards compatible, we only need to set the `out_indices` argument equal to the values of `segmentation_indices`

Comment (amyeroberts) [reply src/transformers/models/beit/configuration_beit.py:207]: We don't just care about backwards compatibility of the original checkpoints. Anyone who's trained a custom beit model will have `out_indices` here e.g. [this model](https://huggingface.co/hkivancoral/hushem_40x_beit_base_f3/blob/main/config.json). 

As mentioned in [my previous comment](https://github.com/huggingface/transformers/pull/25952/#discussion_r1370656949) - this also has unpredicatble behaviour because of how `out_indices` and `out_features` are updated in the config. This still needs to be safely handled.

Comment (amyeroberts) [reply src/transformers/models/beit/configuration_beit.py:207]: As mentioned on slack, this breaks backwards compatibility for the SemanticSegmentation models: 

```
config = BeitConfig()
model = BeitForSemanticSegmentation(config)
```

produces a different model with this change.

File: src/transformers/models/beit/modeling_beit.py [hunk 6]
```diff
@@ -1280,3 +1302,120 @@ def forward(
...
+        >>> import torch
+        >>> from PIL import Image
+        >>> import requests
+
+        >>> url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+        >>> image = Image.open(requests.get(url, stream=True).raw)
+
+        >>> processor = AutoImageProcessor.from_pretrained("microsoft/beit-base-patch16-224")
+        >>> model = AutoBackbone.from_pretrained(
+        ...     "microsoft/resnet-50", out_features=["stage1", "stage2", "stage3", "stage4"]
```
Comment (amyeroberts) [inline src/transformers/models/beit/modeling_beit.py:1366]: A beit checkpoint should be used here

File: src/transformers/models/beit/modeling_beit.py [hunk 7]
```diff
@@ -1151,6 +1167,12 @@ def __init__(self, config: BeitConfig) -> None:
         self.beit = BeitModel(config, add_pooling_layer=False)
 
         # FPNs
+        if len(self.config.out_indices) != 4:
```
Comment (amyeroberts) [inline src/transformers/models/beit/modeling_beit.py:1170]: This error is now going to be raised for everyone trying to create a semantic segmentation model from a default config

File: src/transformers/models/beit/modeling_beit.py [hunk 8]
```diff
@@ -1280,3 +1302,120 @@ def forward(
...
+                feature_maps += (hidden_state,)
+
+        if self.config.add_fpn:
+            feature_maps = [
+                self.fpn1(feature_maps[0]),
+                self.fpn2(feature_maps[1]),
+                self.fpn3(feature_maps[2]),
+                self.fpn4(feature_maps[3]),
+            ]
+            feature_maps = tuple(feature_maps)
```
Comment (amyeroberts) [inline src/transformers/models/beit/modeling_beit.py:1408]: If this is the case then the equivalent check on the number of out_indices also needs to be applied to this model's init.

**Review by amyeroberts — COMMENTED** [review]
Thank you for reworking the docstrings to remove conditional meanings for the parameters. 

I don't like the breaking change in terms of the initialisation of the SemanticSegmentation model. The current warnings help - however it's still subject to unexpected behaviours: if the default `out_indices` or `out_features` changed then this could fail silently. I realise this is a consequence of this being an old model which had param names before we introduced backbones - however the handling `segmentation_indices` highlights other changes and the difficulty in updating config args. 

It's defensive enough that I'll accept once the rest of the comments have been addressed.

File: tests/models/beit/test_modeling_beit.py [hunk 9]
```diff
@@ -64,7 +67,7 @@ def __init__(
         is_training=True,
         use_labels=True,
         hidden_size=32,
-        num_hidden_layers=2,
+        num_hidden_layers=4,
```
[the remark this answers is not in the document]
Comment (NielsRogge) [reply tests/models/beit/test_modeling_beit.py:70]: The semantic segmentation head has a lot of hardcoded values which assume 4 backbone feature maps, such as [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/beit/modeling_beit.py#L1153-L1164) and [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/beit/modeling_beit.py#L1008). Hence if you only use 2 hidden layers, the model will complain.

The reason this still passed on main was because in the current implementation, `hidden_states` also includes the initial embeddings, hence `hidden_states` will be a list of length 3. I see that if I set `num_hidden_layers=3`, the tests also pass.

However, that means we would have to remove the `ValueError` which is raised which notifies the user that one should use 4 feature maps for the backbone, since the head assumes this.

File: src/transformers/models/beit/modeling_beit.py (hunk 7 above)
Comment (NielsRogge) [reply src/transformers/models/beit/modeling_beit.py:1170]: Yes, see my comment above regarding whether we want this or not

File: src/transformers/models/beit/modeling_beit.py (hunk 8 above)
Comment (NielsRogge) [reply src/transformers/models/beit/modeling_beit.py:1408]: Agree. This is also an example of the hardcoded FPNs which assume 4 feature maps (which is why the `ValueError` might make sense)

File: tests/models/beit/test_modeling_beit.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (amyeroberts) [reply tests/models/beit/test_modeling_beit.py:70]: Agreed - let's keep it as 4 here :)

[the remark this answers is not in the document]
Comment (NielsRogge) [reply tests/models/beit/test_modeling_beit.py:70]: The reason it passes with num_hidden_layers=3 is because in that case the first 3 FPNs will be taken [here](https://github.com/huggingface/transformers/blob/main/src/transformers/models/beit/modeling_beit.py#L1254-L1255). There are 3 lateral convs added here, so this allows the model to still perform a forward pass. 3 is basically the minimum of feature maps you need to get it to work.

**Comment by NielsRogge** [comment]
I've added the conversion script to convert DPT + BEiT checkpoints (which is known as DPT 3.1), as BeitBackbone now allows this. Was initially part of #25799

**Comment by NielsRogge** [comment]
@amyeroberts all comments are addressed, thanks for the review!

File: src/transformers/models/beit/modeling_beit.py [hunk 10]
```diff
@@ -1280,3 +1302,126 @@ def forward(
...
+        super()._init_backbone(config)
+
+        self.num_features = [config.hidden_size for _ in range(config.num_hidden_layers + 1)]
+        self.embeddings = BeitEmbeddings(config)
+        self.encoder = BeitEncoder(config, window_size=self.embeddings.patch_embeddings.patch_shape)
+
+        if config.add_fpn:
+            if len(self.config.out_indices) != 4:
+                raise ValueError(
+                    "BeitForSemanticSegmentation requires config.out_indices to be a list of 4 integers, "
```
Comment (amyeroberts) [inline src/transformers/models/beit/modeling_beit.py:1325]:
```suggestion
                    "BeitBackbone requires config.out_indices to be a list of 4 integers, "
```

File: src/transformers/models/beit/modeling_beit.py [hunk 11]
```diff
@@ -1280,3 +1302,126 @@ def forward(
...
+                    "specifying which features to use from the backbone. One can use [3, 5, 7, 11] in case of "
+                    "a base-sized architecture."
+                )
+            hidden_size = config.hidden_size
+            self.fpn1 = nn.Sequential(
+                nn.ConvTranspose2d(hidden_size, hidden_size, kernel_size=2, stride=2),
+                nn.BatchNorm2d(hidden_size),
```
Comment (amyeroberts) [inline src/transformers/models/beit/modeling_beit.py:1332]:
```suggestion
                nn.BatchNorm2d(hidden_size, eps=config.batch_norm_eps),
```

File: src/transformers/models/beit/modeling_beit.py [hunk 12]
```diff
@@ -1280,3 +1302,126 @@ def forward(
...
+
+        self.num_features = [config.hidden_size for _ in range(config.num_hidden_layers + 1)]
+        self.embeddings = BeitEmbeddings(config)
+        self.encoder = BeitEncoder(config, window_size=self.embeddings.patch_embeddings.patch_shape)
+
+        if config.add_fpn:
+            if len(self.config.out_indices) != 4:
+                raise ValueError(
+                    "BeitForSemanticSegmentation requires config.out_indices to be a list of 4 integers, "
+                    "specifying which features to use from the backbone. One can use [3, 5, 7, 11] in case of "
```
Comment (amyeroberts) [inline src/transformers/models/beit/modeling_beit.py:1326]: Is [3, 5, 7, 11] the default we want here?

**Review by amyeroberts — APPROVED** [review]
Thanks for iterating! 

Some small nits just before merging

File: src/transformers/models/beit/modeling_beit.py (hunk 12 above)
Comment (NielsRogge) [reply src/transformers/models/beit/modeling_beit.py:1326]: Yes, similar to the semantic segmentation model, these are good defaults
